### PR TITLE
parser: handle scan identifier meets \0

### PR DESCRIPTION
**Pull request by tiancaiamao**
this is a vague corner case
it should be illegal according to mysql document
but mysql implementation seems accept it
so we follow implementation

fix https://github.com/pingcap/tidb/issues/1989

### CODE REVIEW

File: parser/lexer.go [hunk 1]
```diff
@@ -516,7 +516,7 @@ func (r *reader) peek() rune {
 	v, w := rune(r.s[r.p.Offset]), 1
 	switch {
 	case v == 0:
-		return unicode.ReplacementChar
+		return v // illegal UTF-8 encoding
```
Comment (coocood) [inline parser/lexer.go]: The width `w` need to be set too? Then we don't need to check this case `v == 0'.

**Comment by coocood** [comment]
LGTM

File: parser/scanner_test.go [hunk 2]
```diff
@@ -14,6 +14,7 @@
 package parser
 
 import (
+	"fmt"
```
Comment (shenli) [inline parser/scanner_test.go]: Add a blank line after this line.

File: server/conn.go [hunk 3]
```diff
@@ -369,6 +369,10 @@ func (cc *clientConn) dispatch(data []byte) error {
 	case mysql.ComQuit:
 		return io.EOF
 	case mysql.ComQuery: // Most frequently used command.
+		// trim last '\0' in EOF string
```
Comment (shenli) [inline server/conn.go]: Add comment here.

File: parser/lexer.go [hunk 4]
```diff
@@ -129,6 +129,9 @@ func (s *Scanner) scan() (tok int, pos Pos, lit string) {
 		ch0 = s.skipWhitespace()
 	}
 	pos = s.r.pos()
+	if s.r.eof() {
```
Comment (shenli) [inline parser/lexer.go]: Add comment or reference here.

File: parser/misc.go [hunk 5]
```diff
@@ -83,6 +83,7 @@ func initTokenFunc(str string, fn func(s *Scanner) (int, Pos, string)) {
 }
 
 func init() {
+	ruleTable.token = invalid
```
Comment (shenli) [inline parser/misc.go]: Add comment here.

**Comment by shenli** [comment]
LGTM

**Pull request merged**